lib/eventsource/eventsource.js

```diff
@@ -0,0 +1,326 @@
+   * @param {string} url
+   * @param {EventSourceInit} [eventSourceInitDict]
+   */
+  constructor (url, eventSourceInitDict) {
+    super()
+
+    webidl.argumentLengthCheck(arguments, 1, { header: 'EventSource constructor' })
+
+    // 1. Let baseURL be this's relevant settings object's API base URL.
+    const baseURL = getGlobalOrigin()
+
+    // 2. Let urlRecord be the result of applying the URL parser to url with baseURL.
+    let urlRecord
+
+    try {
+      urlRecord = new URL(url, baseURL)
+    } catch (e) {
+      // 3. If urlRecord is failure, then throw a "SyntaxError" DOMException.
+      throw new DOMException(e, 'SyntaxError')
+    }
+
+    // 4. Set this's url to urlRecord.
+    this.#url = urlRecord.href
+
+    this[kState] = {
+      lastEventId: '',
+      origin: '',
+      reconnectionTime: defaultReconnectionTime
+    }
+
+    this[kEvents] = {
+      message: null,
+      error: null,
+      open: null
+    }
+
+    this[kState].origin = new URL(this.#url).origin
+
+    if (eventSourceInitDict) {
+      if (eventSourceInitDict.withCredentials) {
+        this.#withCredentials = eventSourceInitDict.withCredentials
+      }
+    }
+
+    this.#connect()
+  }
+
+  /**
+   * Returns the state of this EventSource object's connection. It can have the
+   * values described below.
+   * @returns {0|1|2}
+   * @readonly
+   */
+  get readyState () {
+    return this.#readyState
+  }
+
+  /**
+   * Returns the URL providing the event stream.
+   * @readonly
+   * @returns {string}
+   */
+  get url () {
+    return this.#url
+  }
+
+  /**
+   * Returns a boolean indicating whether the EventSource object was
+   * instantiated with CORS credentials set (true), or not (false, the default).
+   */
+  get withCredentials () {
+    return this.#withCredentials
+  }
+
+  async #connect () {
+    this.#readyState = CONNECTING
+    this.#connection = null
+
+    /**
+     * @type {RequestInit}
+     */
+    const options = {
+      method: 'GET',
+      redirect: 'manual',
+      keepalive: true,
+      headers: {
+        Accept: mimeType,
+        'Cache-Control': 'no-cache',
+        Connection: 'keep-alive'
+      },
+      signal: this.#controller.signal
+    }
+
+    if (this.#lastEventId) {
+      options.headers['Last-Event-ID'] = this.#lastEventId
+    }
+
+    options.credentials = this.#withCredentials ? 'include' : 'omit'
+
+    try {
+      this.#connection = await fetch(this.#url, options)
+
+      // Handle HTTP redirects
+      // https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events-intro
+      switch (this.#connection.status) {
+        // Redirecting status codes
+        case 301: // 301 Moved Permanently
+        case 302: // 302 Found
+        case 307: // 307 Temporary Redirect
+        case 308: // 308 Permanent Redirect
+          if (!this.#connection.headers.has('Location')) {
+            this.close()
+            this.dispatchEvent(new Event('error'))
```

Comment:
is this meant to use ErrorEvent? If so, we have an implementation with the other WebSocket events